tests/test-dladdr.c

```diff
@@ -0,0 +1,177 @@
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _MSC_VER
+#include <windows.h>
+#elif _WIN32
+/* for the following functions force a specific import type regardless of the include file in which they are defined */
+#define HMODULE void*
+#define HANDLE void*
+#define LPCVOID void*
+#define LPCSTR char*
+#define DWORD long
+#define SIZE_T long
+
+/* Link by import thunk */
+HMODULE __stdcall GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T __stdcall VirtualQuery (LPCVOID lpAddress, LPCVOID lpBuffer, SIZE_T dwLength);
+/* Link by iat */
+__declspec(dllimport) HMODULE __stdcall LoadLibraryExA (LPCSTR lpLibFileName, HANDLE hFile, DWORD dwFlags);
+#endif
+
+#ifdef _WIN32
+#define PassWithoutSymbolOnWin PassWithoutSymbol
+#define PassWithDifferentAddressOnWin PassWithDifferentAddress
+#else
+#define PassWithoutSymbolOnWin Pass
+#define PassWithDifferentAddressOnWin Pass
+#endif
+
+#ifdef DLFCN_WIN32_SHARED
+#define PassWithSymbolOnSharedBuild Pass
+#else
+#define PassWithSymbolOnSharedBuild PassWithoutSymbol
+#endif
+
+int main(int argc, char **argv)
+{
+    int  result = 0;
+    if (argc == 2)
+        verbose = 1;
+
+    /* null pointer */
+    result = check_dladdr((void*)0, "null" , NoInfo);
+    /* invalid pointer */
+    result |= check_dladdr((void*)0x125, "invalid pointer" , NoInfo);
+    /* -ldl */
+    result |= check_dladdr( ((void*)dladdr), "dladdr" , PassWithSymbolOnSharedBuild );
+    /* -ldl */
+    result |= check_dladdr( (void*)dlopen, "dlopen", PassWithSymbolOnSharedBuild );
+    /* -lglibc */
+    result |= check_dladdr( (void*)atoi, "atoi", PassWithDifferentAddressOnWin );
+    /* test-dladdr */
+    result |= check_dladdr( (void*)main, "main", PassWithoutSymbolOnWin );
+    /* offsets */
+    result |= check_dladdr( (char*)dladdr-6, "dladdr-6", Fail );
+    /* offsets */
+    result |= check_dladdr( (char*)dladdr+6, "dladdr+6", Fail );
+    /* invalid address on import thunk */
+    unsigned char buffer[6] = { 0xFF, 0x25, 0x00, 0x00, 0x00, 0x00 };
+    result |= check_dladdr( buffer, "invalid import thunk", NoInfo );
+
+#ifdef _WIN32
+    /* last entry in iat */
+    result |= check_dladdr( (char*)VirtualQuery, "VirtualQuery", PassWithDifferentAddress );
+    /* links to import thunk table */
+    result |= check_dladdr ( (void*)GetModuleHandleA, "GetModuleHandleA", PassWithDifferentAddress );
+    result |= check_dladdr_by_dlopen( "kernel32.dll", "GetModuleHandleA", Pass );
+
+    /* links directly to Import allocation table */
+    result |= check_dladdr ( (void*)LoadLibraryExA, "LoadLibraryExA", Pass );
+    result |= check_dladdr_by_dlopen( "kernel32.dll", "LoadLibraryExA", Pass );
+#endif
```

Comment:
And... how it is possible that these tests pass on MSVC when code for explicit linking via iat and via import thunk table is not enabled (or rather not working) on MSVC? Should not tests fail on MSVC?

Reply:
By default win api related functions are decorated with  WINBASEAPI (which is defined as __declspec(dllimport)).
Therefore these functions are linked by iat.

Reply:
Ok, this explains why `result |= check_dladdr ( (void*)LoadLibraryExA, "LoadLibraryExA", Pass );` test passed.

But it does not explain why `result |= check_dladdr ( (void*)GetModuleHandleA, "GetModuleHandleA", PassWithDifferentAddress );` test has not failed as it should test import thunk table and therefore have different address.

Reply:
> But it does not explain why result |= check_dladdr ( (void*)GetModuleHandleA, "GetModuleHandleA", PassWithDifferentAddress ); test has not failed as it should test import thunk table and therefore have different address.

check_dladdr() ignores the address completly, if PassWithDifferentAddress has been specified. I fixed that.